tree_assume/src/type_analysis.egg

```diff
@@ -0,0 +1,100 @@
+(ruleset type-analysis)
+
+(relation HasType (Expr Type))
+
+; Primitives
+(rule ((= lhs (Const (Int i))))
+      ((HasType lhs (Base (IntT))))
+      :ruleset type-analysis)
+
+(rule ((= lhs (Const (Bool b))))
+      ((HasType lhs (Base (BoolT))))
+      :ruleset type-analysis)
+
+(rule ((= lhs (Empty)))
+      ((HasType lhs (TupleT (TNil))))
+      :ruleset type-analysis)
+
+; Unary Ops
+(rule (
```

Comment:
I have an idea for how to refactor this rule to cover more erroring cases. How about something like:
```
(rule ((= lhs (Uop (Not) e)))
        ((HasType e (Base (BoolT)))
         (HasType lhs (Base (BoolT)))))
```
This enforces both that the input has type bool and that the output does.
So you'll get a type error for this program: `(Uop (Not) (Int 2))`
Which you won't get an error for right now I think